ui/src/message_popup/scaffolding_card.jsx

```diff
@@ -1,61 +1,149 @@
+      sessionLength: this.state.sessionLength,
+      questions: this.getQuestions(this.state.competencyGroupValue)
+    };
+    this.props.saveScaffoldingAndSession(scaffolding, gameSession);
+  },
   
+  onCompetencyGroupChanged(event, competencyGroupValue){
+    const questions = this.getQuestions(competencyGroupValue);
+    const newLength = questions.length;
+    var sessionLength = this.state.sessionLength;
+    if(sessionLength > newLength){
+      sessionLength = newLength;
+    }
+    this.setState({questions, sessionLength, competencyGroupValue});
+  },
+  
+  onSliderChange(event, value){
+    this.setState({ sessionLength: value });
+  },
   
-  onHelpToggled(){
-    if(this.props.helpType === 'feedback'){
-      this.props.onHelpToggled(null, 'none');
+  onStudentCardsToggled(){
+    this.setState({ shouldShowStudentCard: !this.state.shouldShowStudentCard });
+  },
+  
+  onSummaryToggled(){ 
+    this.setState({ shouldShowSummary: !this.state.shouldShowSummary });
+  },
+  
+  onHelpToggled(event, value){
+    if (typeof value === 'boolean'){
+      this.setState({ helpType: this.state.helpType === 'feedback' ? 'none' : 'feedback'});
     }else{
-      this.props.onHelpToggled(null, 'feedback');
+      this.setState({ helpType: value});
     }
   },
+
+  onTextChanged({target:{value}}:TextChangeEvent) {
+    this.setState({email: value});
+  },
   
   render(){
-    const {competencyGroupValue, competencyGroups, onCompetencyGroupChanged, sessionLength, questions, onSliderChange, shouldShowStudentCard, onStudentCardsToggled, shouldShowSummary, onSummaryToggled, helpType, onHelpToggled, itemsToShow} = this.props;
+    const {itemsToShow} = this.props;
+    const {competencyGroupValue, sessionLength, questions, shouldShowStudentCard, shouldShowSummary, helpType} = this.state;
+
+    const competencyGroups = _.uniq(_.map(allQuestions, 'learningObjectiveId')).map((id) => {
+      return _.find(learningObjectives, {id}).competencyGroup;
+    });
     
-    var showLearningObjectives = true;
-    var showSlider = true;
-    var showStudentCardsToggle = _.has(itemsToShow, 'all') || _.has(itemsToShow, 'cards') || _.has(itemsToShow, 'basic');
-    var showSummaryToggle = _.has(itemsToShow, 'all') || _.has(itemsToShow, 'summary');
-    var showHelpToggle = _.has(itemsToShow, 'all') || _.has(itemsToShow, 'feedback') || _.has(itemsToShow, 'basic');
-    var showOriginalHelp = _.has(itemsToShow, 'all') || _.has(itemsToShow, 'originalHelp');
+    const showLearningObjectives = true;
+    const showSlider = true;
+    const showStudentCardsToggle = _.has(itemsToShow, 'all') || _.has(itemsToShow, 'cards') || _.has(itemsToShow, 'basic');
+    const showSummaryToggle = _.has(itemsToShow, 'all') || _.has(itemsToShow, 'summary');
+    const showHelpToggle = _.has(itemsToShow, 'all') || _.has(itemsToShow, 'feedback') || _.has(itemsToShow, 'basic');
+    const showOriginalHelp = _.has(itemsToShow, 'all') || _.has(itemsToShow, 'originalHelp');
```

Comment:
This looks like something we could simplify, but that's something for another day :)